File: 224.BasicCalc/basicCalc.ts
function calculate(s: string): number {
  const stack: number[] = [];
  let res = 0;
  let num = 0;
  let sign = 1;
  let i = 0;
  while (i < s.length) {
    const char = s[i];
    if (char >= "0" && char <= "9") {
      num = num * 10 + Number(char);
    } else if (char === "+") {
      res += sign * num;
      num = 0;
      sign = 1;
    } else if (char === "-") {
      res += sign * num;
      num = 0;
      sign = -1;
    } else if (char === "(") {
      stack.push(res);
      stack.push(sign);
      res = 0;
      sign = 1;
    } else if (char === ")") {
      res += sign * num;
      num = 0;
      res *= stack.pop()!;
      res += stack.pop()!;
    }
    i++;
  }
  res += sign * num;
  return res;
}

File: 224.BasicCalc/basicCalc.py
class Solution:
    def calculate(self, s: str) -> int:
        stack = []
        res = 0
        num = 0
        sign = 1
        i = 0
        while i < len(s):
            char = s[i]
            if char.isdigit():
                num = num * 10 + int(char)
            elif char == '+':
                res += sign * num
                num = 0
                sign = 1
            elif char == '-':
                res += sign * num
                num = 0
                sign = -1
            elif char == '(':  # Push result and sign for new sub-expression
                stack.append(res)
                stack.append(sign)
                res = 0
                sign = 1
            elif char == ')':
                res += sign * num
                num = 0
                res *= stack.pop()  # sign before '('
                res += stack.pop()  # result calculated before '('
            i += 1
        res += sign * num
        return res